### ISSUE TITLE
System or component identifiers on stereotypes

### ISSUE THREAD

**Issue by romkavt** [plantuml-stdlib/C4-PlantUML#200]
I'd like to discuss a new feature.
I piloting C4 diagrams for enterprise architecture modelling. The typical enterprise has a bunch of large and/or small IT systems.
Every system on corporate IT landscape has it's own identifier enumerated in ITSM database.
I'd like to show these identifiers on context or container diagrams as separate tag or field.
This will be very useful to connect the C4 diagrams repository with ITSM, CMDB and other accounting entities.
What do you think?

**Comment by vitalyzhakov** [comment]
Hello!

Using external directory (catalogs, may be hierarchical) will be useful in designing big systems.

May be this feature implemented by generation file for including with entities?
Can it solve issue?

**Comment by kirchsth** [comment]
C4-PlantUML supports [Element properties](https://github.com/plantuml-stdlib/C4-PlantUML/blob/master/README.md#element-properties). Could this be a solution of the problem?

If it basically ok, we could try a better layout (suggestions are welcome)

BR Helmut

![](https://www.plantuml.com/plantuml/png/ZP9HRzCm4CVVyobCNaYbhc4L4X9FcpeGI6Mhe834ItLDZiQIuxFiiqP0V7VEqRfiQO0z-NB--P_xnRa839vZQx9dsbOcrgWQPXTUbwM7syL1SnFtCQ2lo39QNbJKbiw0JMVE0jT6xylLoxDDQdt-i2vR28nUMhihT8QwDXrowGNPSrNZTuY6LODGerSRJmuzTtFr1Kp4xBAkZwqYluOMyxdAtne8JJvxl7dZ3s3rJs1DDa7VY9YSXZ6t9J9f_xrbz1PPlVaXGtdqwjNYXS0Rz85iuVhbqcW80gzXZ_sf6vVomQWh39NN_PCgRZKtzoRkxbLtIZF9p3uX7oTurtUB_FYSp_Easeiz21sFdQhpnFImL8bcq2QSJw7BUtJv05qAEjp1xffgtAqBAylVHRUTm_-OLp4mjHFYwbUMAVLL68hZ3p2JdPEnLuEYbDF8e2PbGbPanSvAPdMiJdIsM3MM31swVxjGdBp0ttA5NM1iYz0lu_od9MeC_T_m4StZ_sjgxb7k82095sZhs9e_)

**Comment by kirchsth** [comment]
In combination with json based data it could look like 

![](https://www.plantuml.com/plantuml/png/ZPFTJXin48Nl_HI36ucWfXirz0ef4e4Yk80MyLFQeWWvws7NgjVUsbv-YlBkzJYcIMY5UfFTFJxpsNCysVL1kD0rccqeK-fE8jGXj7vdE7JYlgXKgBjPvz6LrWGqeIXjC_oXN5dxK0_tFxmxrSA4ovFZ8Jx4TOaoYZVxjccj8NjBQ3QHyXITj-P2L7qUK3GE9Hz0RrRjMstTcB_9WiOgKYuY3CQymMQ63kmjK0cGPUiVh9dkS1sMbLTfv8aujj6J8eZn4szZ41eDtm4-YiMy4WtIns5yKfUK4kdaKwLge-BZ-ZLxuYPR7nv_SHXmGJh_-kqxsMap-foasdJoLCOdxqcPZDBlDf_2VB22SZuiKMVd5wkelPUelOJQ6t6Oha6wk_ibvFBBrUkGKO9i4sGwOBoaCEFA4v-GdzAdTNfFsStPd36s2LXK1KZbMoqUeNKsCeD23_SgrBOBKAEGwDZLulDqONWyIcgVBk3MEXHb3JrA5fI1T2F5QiuC03P1An_e_eJMuF1dXpwWNDcIN3GUUOi-LtJaX2qeadXyYRvqV8kSVuQql-B8Rk5JKcLjadTvn3n4ocHZ3Z1wegj7sSQnX76E-UsYavJazQA0z6co_tzFwqsTFCynzPPttFhRDZ4srEWIAbrrNe6YA6ZkgM3nH-hdhWQmV0cYyZ1-mJz3lU9TZ_KEjMqNOQ6HCI_s_6JdHv-lReuF3Wy-VOptlHlbhj6_0G00)

```plantuml
@startuml
!include https://raw.githubusercontent.com/kirchsth/C4-PlantUML/extended/C4_Component.puml

AddPersonTag("teamred", $bgColor="#red", $legendText ="member of team red")
AddPersonTag("teamblue", $bgColor="#blue", $legendText ="member of team blue")

!$data={"partlen": "2", "game": "GamePlantuml", "participants": [
{"name": "xyz" ,"as": "XYZ", "team": "A", "persons": ["A1", "A2", "A3"] },
{"name": "rst" ,"as": "RST", "team": "B", "persons": ["B1", "B2" ]},
{"name": "uvw" ,"as": "UVW", "team": "B", "persons": ["B2", "B3" ]}],
"color": {"A": "red", "B": "blue"}
}


' e.g. display properties without header
WithoutPropertyHeader()

!foreach $part in $data.participants
   ' list of all requested properties
   AddProperty("id", $part.name, "descr")
   Component($part.name, $part.as, "techn")
  !foreach $member in $part.persons
    !$teamTag = "team"+$data.color[$part.team]
    
   ' list of all requested properties
    AddProperty("Member of", $teamTag)
    AddProperty("Another prop", "property ...")

    Person($member, $member, $tags = $teamTag)
    Rel($member, $part.name, "develop")
  !endfor
!endfor

SHOW_LEGEND()
@endumll
```

**Comment by romkavt** [comment]
Thanks for your suggestions.
I've solved my issue by using hyperlinks to ITSM records.
Here is an example:
The whole system is divided to separate repos by product verticals.
Each repo contains the map of the tech components involved into this solution.
The repo contains two files:
1. the list of tech components with records like this:
```
!procedure PublicCardApi()
Container(PublicCardApi, "public-card-api", "Java", "API для интеграций ИС партнеров",$link="https://jira.yooteam.ru/browse/COM-1531")
!endprocedure
```
2. the solution scheme file:
```
Enterprise_Boundary(YM, "YooMoney") {

    Boundary(PciDss, "PCI-DSS периметр") {
...
        CardApi()
        PublicCardApi()

        Rel(AppWallet, CardApi, "Выпуск и управление картой")
        Rel(AppSdkCards, CardApi, "Выпуск и управление картой")
....
```

Also, we are using System of Interest (blue) / External systems (gray) notation to divide tech systems implementing the solution and enabling systems provided by external teams.

As the result: we have chosen the hypertext environment to integrate С4 as navigation map to EA landscape repository.

**Comment by romkavt** [comment]
And hyperlinked symbols looks like this. I suggest to adjust the look and feel style, making hyperlinks only on title

![изображение](https://user-images.githubusercontent.com/5063381/151947145-ece0a2d0-623f-4942-8fdc-0d136b33af1f.png)

**Comment by kirchsth** [comment]
I think the underline is a related to your used CSS styles. On the plantuml server and its genrated svg files there are no links displayed with an underscore at all

http://www.plantuml.com/plantuml/uml/hT11IpCn40RWUtv5lkKxRC5sBvu4eT8UgnQivv9cnsumcGcP2MKH_xlPbI8YMi6Utic4bpnapcAIvE3LFuVMvnPq9nBvgccIESpsJhgyomp94WgWp2o59dg3GsVAqdgtQnQNq_MmUhnTDS6mG2ghxHAYfpuCfJWyeDQGcB2EOznuP_X2Lokp0r-LuJwA8pHUBu5jSkD9ruS2q2qycUn5nqIbBBtkmBIG9gNstpjylgwEvdVlAMGrKUf3jxsZ5hQh6ajyhzhq4KumRMQXS4QbVb5QBmZ5E8HKs-FqHRa1so5vslS_4ocVzHDVbNheMI3KFCOdrwyb3lWlY3bWMoxV0000

(you have to click on the image that the svg is opened, and then you can follow the links)
![](http://www.plantuml.com/plantuml/svg/hT11IpCn40RWUtv5lkKxRC5sBvu4eT8UgnQivv9cnsumcGcP2MKH_xlPbI8YMi6Utic4bpnapcAIvE3LFuVMvnPq9nBvgccIESpsJhgyomp94WgWp2o59dg3GsVAqdgtQnQNq_MmUhnTDS6mG2ghxHAYfpuCfJWyeDQGcB2EOznuP_X2Lokp0r-LuJwA8pHUBu5jSkD9ruS2q2qycUn5nqIbBBtkmBIG9gNstpjylgwEvdVlAMGrKUf3jxsZ5hQh6ajyhzhq4KumRMQXS4QbVb5QBmZ5E8HKs-FqHRa1so5vslS_4ocVzHDVbNheMI3KFCOdrwyb3lWlY3bWMoxV0000)

Maybe you can add a new style in combination with <text font-weight="bold"> and that only this part is displayed with an underscore (but I'm not a CSS expert)

**Comment by Potherca** [comment]
> Maybe you can add a new style in combination with <text font-weight="bold"> and that only this part is displayed with an underscore (but I'm not a CSS expert)

That would be:

```
a > text[font-weight="bold"]:nth-child(3),
a > text[font-weight="bold"]:nth-child(4) {
    fill: fuchsia;
    text-decoration: underline 4px;
}
```

![](https://raw.githubusercontent.com/gist/Potherca/1fa9fe8b46824f0c6ceaa829ff586bdd/raw/e647b44823ff043af90d0f14524556f9a70d3e01/example.svg)

(3 for labels without icon, 4 for labels with icon)

**Comment by kirchsth** [comment]
@Potherca, 
what is your planed enhancement in this area?
BR Helmut

**Comment by Potherca** [comment]
Unless someone has a concrete suggestion, I don't see any hard need for changing / adding anything for this, as both @romkavt and @vitalyzhakov's use-cases seem to be covered using existing features...

**Comment by kirchsth** [comment]
Ok thanks, I remove the enhancement and not-stale labels

**Comment by romkavt** [comment]
Thanks for suggestions!
We are using PlantUML embedded into Asciidoctor documents as SVG image.
The quick fix solution is: add the
`g a {
    text-decoration: none;
}`
to the CSS stylesheet for asciidoctor document.

But, let's look deep in the code, for exxample https://github.com/plantuml-stdlib/C4-PlantUML/blob/master/C4_Context.puml#L253
`!unquoted procedure System($alias, $label, $descr="", $sprite="", $tags="", $link="")
!$sprite=$toElementArg($sprite, $tags, "ElementTagSprite", "system")
rectangle "$getSystem($label, $descr, $sprite)$getProps()" $toStereos("system", $tags) as $alias $getLink($link)
!endprocedure`

Looks like the hyperlink adds to the whole block, not to it's title only.
That's the problem: asciidoctor treats whole block as HTML hyperlink.

**Comment by romkavt** [comment]
As mentioned in original topic, I have to found the solution how to classify tech systems.
I've found the solution. That's the example case.

First, the C4 notation follows systems theory and this is great!
The whole system has been divided to a couple of functional contexts, mostly by end-user products.
The tech component schemes inside functional context uses blue symbols for tech components inside the context (system of interest) and grey symbols for external enabling systems.

The second thing is Tags. I've added tags to symbols, reflecting the type of the system (end product, enabling platform,  backoffice operations support). Then assigned colors.

There are some examples.

Component diagram:
![изображение](https://user-images.githubusercontent.com/5063381/165318217-c8da87fd-df5b-465d-96ed-90d49d78a851.png)

Context diagram:
![изображение](https://user-images.githubusercontent.com/5063381/165318446-eb7d1bc6-5b36-4b05-a4e1-d77ddb6af3a5.png)
